### DISCUSSION TITLE
Max Number of WebSocket Connections that a Single Nats Server can Sustain Reliably

### DISCUSSION THREAD

**Discussion by dsidirop** [nats-io/nats-server#2770] [Q&A] [answered]
( Assume below that the nats-server in question has 128GB of RAM and "beast-mode" ARM CPUs )

1. Are there any rules of thumb documented somewhere for teams interested in figuring out what is the max-client limit for a core-nats-server when said clients connect to the nats-server using **WebSocket** connections (no leaf-nodes involved) with a specific network-usage profile?

2. For example if on average every client is expected to transfer 1000 messages/hour with message size ~= 1kb every single day, then how many WebSocket connections can a single nats-server sustain without facing issues? (connections getting dropped etc)

3. Does the resulting max-number change if JetStream is involved? Or does the "formula" remain the same regardless?

4. Does the resulting max-number change if WebTransport is used later on (upgrade from WebSocket)? Or does the "formula" remain the same regardless?

PS: I'm sorry if this has been answered elsewhere but I just couldn't spot anything anywhere in the official docs

**Comment by derekcollison** [comment, marked as answer]
This is a great question.

In general, I usually start with what resource will be most constrained. This can be memory, network, or CPU (depending on latency and throughput targets).

For connection's memory footprint, it's fairly simple to run a quick test to measure for yourself the weight of an idle connection in your setup. Last we looked you should see ~20-30k or so per on avg.

Note to pay attention to how many connections would need to fail over in case of a server loss or upgrade, and whether those connections are TLS based (esp RSA). That could cause thundering herd and CPU issues. Most clients have jitter baked in, and ECC keys help alot, but something to keep in mind when scaling up the connections per host. I have ran experiments with ~1M mostly idle connections on a single server.

For CPU it usually is around a throughput or latency target, meaning the network is not the constrained resource.

For the situation above, since you have a good sense of traffic patterns, we can more accurately look at the network resources. If I am doing my math correct, each client does ~1mb/hr or ~8mbit/hr, which translates to 2222.22 bps. Good number ;)

So again, if math correct, that means on a 10Gb network interface, you can do ~4M connections.

So for above I would say your main concerns will be memory (you have plenty in the above scenario) and client reconnects on server disruption as what to plan around. More the later then the former in my opinion.

Hope that helps.

**Reply by dsidirop** [reply]
Amazing! Hats off!
